File: 149-max-points-on-a-line/max-points-on-a-line.py
from fractions import Fraction

class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        if len(points) <= 1:
            return len(points)

        max_points = 1

        for i in range(len(points)):
            slopes = {}

            for j in range(len(points)):
                if i == j:
                    continue

                dx = points[j][0] - points[i][0]
                dy = points[j][1] - points[i][1]

                if dx == 0:
                    slope = float('inf')
                else:
                    slope = Fraction(dy, dx)

                slopes[slope] = slopes.get(slope, 0) + 1

            print(slopes)  # 추가된 부분

            current_max = max(slopes.values() or [0]) + 1
            max_points = max(max_points, current_max)

        return max_points

File: 149-max-points-on-a-line/max-points-on-a-line.java
class Solution {
    public int maxPoints(int[][] points) {
        if (points.length <= 1) {
            return points.length;
        }

        int maxPoints = 1;

        for (int i = 0; i < points.length; i++) {
            HashMap<Double, Integer> slopes = new HashMap<>();

            for (int j = 0; j < points.length; j++) {
                if (i == j) {
                    continue;
                }

                double slope;
                if (points[i][0] == points[j][0]) {
                    slope = Double.POSITIVE_INFINITY;  // 수직인 경우 무한대로 표현
                } else {
                    slope = (double) (points[j][1] - points[i][1]) / (points[j][0] - points[i][0]);
                }

                slopes.put(slope, slopes.getOrDefault(slope, 0) + 1);
            }

            maxPoints = Math.max(maxPoints, slopes.values().stream().max(Integer::compareTo).orElse(0) + 1);
        }

        return maxPoints;
    }
}